src/demangle.cpp

```diff
@@ -25,80 +25,23 @@ char* Demangle::demangleCpp(const char* s) {
     return result;
 }
 
-char* Demangle::demangleRust(const char* s, const char* e) {
+char* Demangle::demangleRust(const char *s, struct demangle const *demangle) {
     // Demangled symbol can be 1.5x longer than original, e.g. 1A1B1C -> A::B::C
-    char* result = (char*)malloc((e - s) * 3 / 2 + 1);
-    if (result == NULL) {
-        return NULL;
-    }
-
-    char* r = result;
-    char* tmp;
-
-    while (s < e) {
-        unsigned long len = strtoul(s, &tmp, 10);
-        const char* next = tmp + len;
-        if (len == 0 || next > e) {
-            break;
-        }
-
-        s = tmp;
-        if (s[0] == '_' && s[1] == '$') s++;
-
-        if (r > result) {
-            *r++ = ':';
-            *r++ = ':';
+    char* result = NULL, *new_result;
+    for (size_t demangled_size = strlen(s) * 3 / 2 + 1; demangled_size < 1000000; demangled_size *= 2) {
+        new_result = (char *)realloc(result, demangled_size);
+        if (new_result == NULL) {
+            free(result);
+            return NULL;
         }
-
-        while (s < next) {
-            if (s[0] == '$') {
-                if (s[1] == 'L' && s[2] == 'T' && s[3] == '$') {
-                    *r++ = '<';
-                    s += 4;
-                } else if (s[1] == 'G' && s[2] == 'T' && s[3] == '$') {
-                    *r++ = '>';
-                    s += 4;
-                } else if (s[1] == 'L' && s[2] == 'P' && s[3] == '$') {
-                    *r++ = '(';
-                    s += 4;
-                } else if (s[1] == 'R' && s[2] == 'P' && s[3] == '$') {
-                    *r++ = ')';
-                    s += 4;
-                } else if (s[1] == 'S' && s[2] == 'P' && s[3] == '$') {
-                    *r++ = '@';
-                    s += 4;
-                } else if (s[1] == 'B' && s[2] == 'P' && s[3] == '$') {
-                    *r++ = '*';
-                    s += 4;
-                } else if (s[1] == 'R' && s[2] == 'F' && s[3] == '$') {
-                    *r++ = '&';
-                    s += 4;
-                } else if (s[1] == 'C' && s[2] == '$') {
-                    *r++ = ',';
-                    s += 3;
-                } else if (s[1] == 'u') {
-                    *r++ = (char)strtoul(s + 2, &tmp, 16);
-                    s = tmp + 1;
-                } else {
-                    *r++ = '$';
-                    s++;
-                }
-            } else if (s[0] == '.' && s[1] == '.') {
-                *r++ = ':';
-                *r++ = ':';
-                s += 2;
-            } else {
-                *r++ = *s++;
-            }
-        }
-
-        if (s > next) {
-            break;
+        result = new_result;
+        if (rust_demangle_display_demangle(demangle, result, demangled_size, true /* alternate */) == OverflowOk) {
+            return result;
         }
     }
-
-    *r = 0;
-    return result;
+    free(result);
+    // demangling Rust failed, just return the original string
+    return strdup(s);
```

Comment:
The function is supposed to return `NULL` if it's not a valid Rust symbol. Then the caller attempts to demangle it as a C++ symbol, and if it also fails, returns the original string.

Reply:
will change